Rules/AvoidExclamationPointOperator.cs

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+#if !CORECLR
+using System.ComponentModel.Composition;
+#endif
+using System.Globalization;
+using System.Linq;
+using System.Management.Automation;
+using System.Management.Automation.Language;
+using Microsoft.Windows.PowerShell.ScriptAnalyzer.Generic;
+
+namespace Microsoft.Windows.PowerShell.ScriptAnalyzer.BuiltinRules
+{
+    /// <summary>
+    /// AvoidExclamationPointOperator: Checks for use of the exclamation point operator
+    /// </summary>
+#if !CORECLR
+    [Export(typeof(IScriptRule))]
+#endif
+    public class AvoidExclamationPointOperator : ConfigurableRule
+    {
+
+        /// <summary>
+        /// Construct an object of AvoidExclamationPointOperator type.
+        /// </summary>
+        public AvoidExclamationPointOperator() {
+            Enable = false;
+        }
+
+        /// <summary>
+        /// Analyzes the given ast to find the [violation]
+        /// </summary>
+        /// <param name="ast">AST to be analyzed. This should be non-null</param>
+        /// <param name="fileName">Name of file that corresponds to the input AST.</param>
+        /// <returns>A an enumerable type containing the violations</returns>
+        public override IEnumerable<DiagnosticRecord> AnalyzeScript(Ast ast, string fileName)
+        {
+            if (ast == null) throw new ArgumentNullException(Strings.NullAstErrorMessage);
+
+            var diagnosticRecords = new List<DiagnosticRecord>();
+
+            IEnumerable<Ast> foundAsts = ast.FindAll(testAst => testAst is UnaryExpressionAst, true);
+            if (foundAsts != null) {
+                var CorrectionDescription = Strings.AvoidExclamationPointOperatorCorrectionDescription;
+                foreach (UnaryExpressionAst foundAst in foundAsts) {
+                    if (foundAst.TokenKind == TokenKind.Exclaim) {
+                        // If the exclaim is not followed by a space, add one
+                        var replaceWith = "-not";
+                        if (foundAst.Child != null && foundAst.Child.Extent.StartColumnNumber == foundAst.Extent.StartColumnNumber + 1) {
```

Comment:
I don't understand what the second part does. as per comment you check if there is a space after '!' but don't see how you achieve that with this comparison, can you help explain please?

Reply:
The intent was that if someone has `!$MyVar`, then instead of the replacement being `-not$MyVar` which looks wrong to my eyes - instead add a space after the `-not` to make the replacement `-not $MyVar`.

The `UnaryExpressionAST` should have a single child, the argument that the unary operator is acting upon. 

Line 52 was checking if that child's extent started at the next column over from the exclaim operator token's extent - implying there is nothing between the Token, `!`, and it's child.

Thinking about it a bit more, I have 2 questions that I could use your advice on:
1. Should the rule care about whitespace - or simply replace appropriate instances of `!` with `-not`? Other rules take care of formatting, as the user prefers, after all.
2. If it should add the whitespace in instances it's missing, is there a different way I should approach doing so?

I appreciate you taking the time to look this over!

Reply:
what you are trying to achieve makes sense, I just didn't understand how you achieved it but the way you explained it with the child ast containing the variable makes sense.